source/slang/reflection.cpp

```diff
@@ -85,6 +95,107 @@ static inline SlangReflection* convert(ProgramLayout* program)
     return (SlangReflection*) program;
 }
 
+// user attaribute
+
+unsigned int getUserAttributeCount(Decl* decl)
+{
+    unsigned int count = 0;
+    for (auto x : decl->GetModifiersOfType<UserDefinedAttribute>())
+    {
+        SLANG_UNUSED(x);
+        count++;
+    }
+    return count;
+}
+
+SlangReflectionUserAttribute* findUserAttributeByName(Session* session, Decl* decl, const char* name)
+{
+    auto nameObj = session->getNameObj(name);
+    for (auto x : decl->GetModifiersOfType<UserDefinedAttribute>())
+    {
+        if (x->name == nameObj)
+            return (SlangReflectionUserAttribute*)(x);
+    }
+    return nullptr;
+}
+
+SlangReflectionUserAttribute* getUserAttributeByIndex(Decl* decl, unsigned int index)
+{
+    unsigned int id = 0;
+    for (auto x : decl->GetModifiersOfType<UserDefinedAttribute>())
+    {
```

Comment:
As mentioned before, this loop is a bit gross because it means that clients of the API might not realize that their "loop over all attributes" code is actually quadratic in the number of attributes. For now I guess it is fine since the number of attributes will in practice always be small.

The obvious fix here is to have `ModifiableSyntaxNode` store an array of modifiers instead of a singly-linked list. I think that is reasonable to do *if* we make a similar change to `List<T>` as was made to `String` to make it a single pointer in size, with copy-on-write semantics. At that point the overhead of adding a full-fledged array to `ModifiableSyntaxNode` would be the same as the current linked list.

Reply:
Although, making the list of modifiers be a real array doesn't exactly fix this case unless we change this to iterate over *all* attributes and not just user-defined ones.

Reply:
Agreed that this is gross. Not sure what is the best way to fix it.